samples/solutions/producer_consumer.py

```diff
@@ -0,0 +1,244 @@
+                    self.log_info_and_print("Successfully connected the device client...")
+                except Exception as e:
+                    if self.first_connect:
+                        self.log_info_and_print(
+                            "Very first connection never occurred so will retry immediately..."
+                        )
+                        self.first_connect = False
+                        sleep_time = 0
+                    else:
+                        self.log_info_and_print(
+                            "Retry attempt interval is {} and increase power factor is {}".format(
+                                self.sleep_time_between_conns, self.retry_increase_factor
+                            )
+                        )
+                        sleep_time = pow(self.sleep_time_between_conns, self.retry_increase_factor)
+
+                    if sleep_time > THRESHOLD_FOR_RETRY_CONNECTION:
+                        self.log_error_and_print(
+                            "Failed to connect the device client couple of times."
+                            "Retry time is greater than upper limit set. Will be exiting the application."
+                        )
+                        self.try_number = 0
+                        raise
+
+                    self.log_error_and_print("Caught exception while trying to connect...")
+                    self.log_error_and_print(
+                        "Failed to connect the device client due to error :{}.Sleeping and retrying after {} seconds".format(
+                            get_type_name(e), sleep_time
+                        )
+                    )
+                    self.retry_increase_factor += 1
+                    self.try_number += 1
+                    await asyncio.sleep(sleep_time)
+
+    def log_error_and_print(self, s):
+        logger.error(s)
+        print(s)
+
+    def log_info_and_print(self, s):
+        logger.info(s)
+        print(s)
+
+    async def main(self):
+        conn_str = os.getenv("IOTHUB_DEVICE_CONNECTION_STRING")
+        await self.initiate()
+        await self.create_client(conn_str)
+
+        tasks = [
+            asyncio.create_task(self.wait_for_connect_and_send_telemetry()),
+            asyncio.create_task(self.enqueue_message()),
+            asyncio.create_task(self.if_disconnected_then_connect_with_retry()),
+        ]
+
+        try:
+            done, pending = await asyncio.wait(tasks, return_when=asyncio.FIRST_EXCEPTION)
+            await asyncio.gather(*done)
+        except KeyboardInterrupt:
+            self.log_error_and_print("IoTHubClient sample stopped by user")
+        except Exception as e:
+            self.log_error_and_print("Exception in main loop: {}".format(get_type_name(e)))
+        finally:
+            self.log_info_and_print("Exiting app")
+            self.exit_app_event.set()
+            self.log_info_and_print("Waiting for all coroutines to exit")
```

Comment:
>self.log_info_and_print("Waiting for all coroutines to exit") [](http://example.com/codeflow?start=12&length=61)

You should decide how you want to shut down these tasks. You set `app_exit_event`, which should eventually cause the tasks to complete, but you don't wait for the tasks to actually exit. If you're just going to cancel all the tasks after setting this event, then having this event in the code doesn't really do much.

Reply:
actually i did both types one to wait and one to cancel it did not really change the behavior - i expected it would. But i have deleted the cancel for now.